src/main/scala/com/microsoft/hyperspace/index/rules/HyperspaceRule.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.rules
+
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+
+import com.microsoft.hyperspace.ActiveSparkSession
+import com.microsoft.hyperspace.index.IndexLogEntry
+
+/**
+ * Interface of exclusive type of indexes.
+ */
+trait HyperspaceRule extends ActiveSparkSession {
+
+  /**
+   * Sequence of conditions to apply indexes. Each filter contains conditions and
+   * filter out candidate indexes based on the conditions. The order of the sequence does matter.
+   */
+  val planFilters: Seq[PlanFilter]
+
+  /**
+   * Transform the plan to use the selected indexes.
+   *
+   * @param plan Original query plan.
+   * @param indexes Selected indexes.
+   * @return Transformed plan to use the selected indexes.
+   */
+  def applyIndex(plan: LogicalPlan, indexes: Map[LogicalPlan, Seq[IndexLogEntry]]): LogicalPlan
+
+  /**
+   * Calculate the score of the index application.
+   *
+   * @param plan Original query plan.
+   * @param indexes Selected indexes.
+   * @return Score of selected indexes.
+   */
+  def score(plan: LogicalPlan, indexes: Map[LogicalPlan, Seq[IndexLogEntry]]): Int
```

Comment:
Question. I see the following below:
```scala
(applyIndex(plan, candidateIndexes), score(plan, candidateIndexes))
```
Are you trying to apply `candidateIndexes` again to plan inside `score`? Don't you need to utilize the result of `applyIndex`?

Reply:
I assume that `candidateIndexes` at the point is the final candidate which means *all* `candidateIndexes` should be "applicable" to the given plan. I split the function for better structure & function role for readability.

Reply:
Shouldn't `score` use the same apply logic as `applyIndex`? Do you have a sample implementation?

If I understand the intention of `score` correctly, it seems that you can inject `Scorer` trait into `applyIndex` instead, and invoke it whenever an index is applied.

Reply:
https://github.com/microsoft/hyperspace/pull/403/files#diff-c3a2b337d86b3cc05c4ac4cdafa48c332221c03b38df5cd6a22e315541334d08R516 - example for JoinIndex

> If I understand the intention of score correctly, it seems that you can inject Scorer trait into applyIndex instead, and invoke it whenever an index is applied.

We might need to consider left & right index together (though the sample doesn't); so I thought it would be better to calculate the score with a different function.

Reply:
I guess my uber question is is there a relationship b/w `applyIndex` and `score` or are they totally orthogonal?

Also, I see the following in the sample code for `score`: `val candidateIndex = indexes.head._2.head`. So we don't care which index is actually applied?

Reply:
Yes they are orthogonal. `applyIndex` should be able to apply those indexes.

Reply:
OK, then what do you mean by `Calculate the score of the index application.`? This index application is different from what's happening inside `applyIndex`?

Reply:
I revised it to "score of the selected indexes". It was because evaluating indexes assumes index application.